### PR TITLE
Workaround to evaluate javascript using UIWebView

### PR DESCRIPTION
**Pull request by jboich**
Workaround for issue [#988 ](https://github.com/Microsoft/WinObjC/issues/988)

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/winobjc/1638)

### CODE REVIEW

File: Frameworks/UIKit/UIWebView.mm [hunk 1]
```diff
@@ -251,6 +251,25 @@ - (NSString*)stringByEvaluatingJavaScriptFromString:(NSString*)string {
 }
 
 /**
+ @Status Caveat
+  @Notes This is a workaround. Original UIWebView does not have this method
+*/
+- (void)evaluateJavaScript:(NSString*)javaScriptString completionHandler:(void (^)(id, NSError*))completionHandler {
```
Comment (rajsesh-msft) [inline Frameworks/UIKit/UIWebView.mm]: Rather than adding a new function, you can do this in stringByEvaluatingJavaScriptFromString.

Comment (jaredhms) [inline Frameworks/UIKit/UIWebView.mm]: stringByEvaluatingJavaScriptFromString isn't async though.

---
In reply to: [95190774](https://github.com/Microsoft/WinObjC/pull/1638#discussion_r95190774) [](ancestors = 95190774)

Comment (rajsesh-msft) [inline Frameworks/UIKit/UIWebView.mm]: ah, yes.  Inexplicably, I thought this function was "waiting" for the eval.  It isn't. Apologies for the false flag.

---
In reply to: [95208519](https://github.com/Microsoft/WinObjC/pull/1638#discussion_r95208519) [](ancestors = 95208519,95190774)

**Comment by jaredhms** [comment]
>     return ret;

Since we know this will never work, maybe we should change this status to Stub and replace this code with:

UNIMPLEMENTED_WITH_MSG("stringByEvaluatingJavaScriptFromString is not supported on our platform.  Call evaluateJavaScript:completionHandler: instead.");
return StubReturn();

?

We'd also need to add STUB_METHOD to the declaration in the header file, possibly with the same comment saying why it's stubbed out:

// stringByEvaluatingJavaScriptFromString is not supported on our platform.  Call evaluateJavaScript:completionHandler: instead.
- (NSString*)stringByEvaluatingJavaScriptFromString:(NSString*)script STUB_METHOD;

---
Refers to: Frameworks/UIKit/UIWebView.mm:250 in 6c9c08e. [](commit_id = 6c9c08ec9a2365ebc0304b80a1fcf8d5f1e2d2b1, deletion_comment = False)

**Comment by jaredhms** [comment]
>     return ret;

@Raj Seshasankaran - this is another place where an @Status Unsupported and a UNSUPPORTED_WITH_MSG would be helpful; agree?  

---
In reply to: [271356945](https://github.com/Microsoft/WinObjC/pull/1638#issuecomment-271356945) [](ancestors = 271356945)

---
Refers to: Frameworks/UIKit/UIWebView.mm:250 in 6c9c08e. [](commit_id = 6c9c08ec9a2365ebc0304b80a1fcf8d5f1e2d2b1, deletion_comment = False)

**Comment by rajsesh-msft** [comment]
>     return ret;

We don't (yet) have a status unsupported.  At the very least, this should be status Gap.

---
In reply to: [271357323](https://github.com/Microsoft/WinObjC/pull/1638#issuecomment-271357323) [](ancestors = 271357323,271356945)

---
Refers to: Frameworks/UIKit/UIWebView.mm:250 in 6c9c08e. [](commit_id = 6c9c08ec9a2365ebc0304b80a1fcf8d5f1e2d2b1, deletion_comment = False)

**Comment by jaredhms** [comment]
>     return ret;

@jboich, can you please use status Gap for now rather than status Stub?

/**
 @Status Gap
*/


---
In reply to: [271357323](https://github.com/Microsoft/WinObjC/pull/1638#issuecomment-271357323) [](ancestors = 271357323,271356945)

---
Refers to: Frameworks/UIKit/UIWebView.mm:250 in 6c9c08e. [](commit_id = 6c9c08ec9a2365ebc0304b80a1fcf8d5f1e2d2b1, deletion_comment = False)

**Comment by jaredhms** [comment]
:shipit:

**Comment by jaredhms** [comment]
Thanks for the fix and updates; kicking off a CI build.

File: Frameworks/UIKit/UIWebView.mm [hunk 2]
```diff
@@ -243,11 +243,31 @@ - (void)setDetectsPhoneNumbers:(BOOL)detect {
...
 /**
- @Status Interoperable
+ @Status Gap
 */
 - (NSString*)stringByEvaluatingJavaScriptFromString:(NSString*)string {
-    NSString* ret = [_xamlWebControl invokeScript:@"eval" arguments:[NSArray arrayWithObject:string]];
-    return ret;
+    UNIMPLEMENTED_WITH_MSG(
+        "stringByEvaluatingJavaScriptFromString is not supported on our platform. Call evaluateJavaScript:completionHandler: instead.");
+    return StubReturn();
```
Comment (jaredhms) [inline Frameworks/UIKit/UIWebView.mm]: This is hitting a build break;

D:\Build\Agent1\_work\2\s\Frameworks\UIKit\UIWebView.mm(251,12): error : use of undeclared identifier 'StubReturn' [D:\Build\Agent1\_work\2\s\build\UIKit\lib\UIKitLib.vcxproj]

Can you please add #include "StubReturn.h" just below the #include "Starboard.h"?

Comment (jboich) [inline Frameworks/UIKit/UIWebView.mm]: Sure, I will add. Sorry, I missed this. Why the build did not break in Debug mode?

Comment (jaredhms) [inline Frameworks/UIKit/UIWebView.mm]: Thanks, no problem.  I'm not sure why it didn't fail for you; it failed on both debug and release CI builds.

---
In reply to: [95416396](https://github.com/Microsoft/WinObjC/pull/1638#discussion_r95416396) [](ancestors = 95416396)

**Pull request merged**